docs(hpa): How HPA works in CDaaS (#1499)
armory/docs: content/en/cd-as-a-service/concepts/deployment/how-hpa-works.md

---

> title: "Kubernetes Horizonal Pod Autoscaler deployments in CD-as-a-Service"

Reviewer: Capital case in titles.

Either `HorizontalPodAutoscaler` or Horizontal Pod Autoscaling

linktitle: "Kubernetes HPA"
description: >
  Learn how Armory Continuous Deployment-as-a-Service supports Kubernetes deployments with Horizontal Pod Autoscaling.
---

# Before you begin

> Read the Kubernetes Documentation topics: 

Reviewer: What do you think about softening this a bit? 
Maybe something like "Make sure you are familiar with using Horizontal Pod Autoscaling in Kubernetes deployments" or "You understand Horizontal Pod Autoscaling in Kubernetes"  or "Before you begin using a `HorizontalPodAutoscaler` in your CD-as-a-Service deployment, you should have a working knowledge of Horizonal Pod Autoscaling in Kubernetes"  and embed a link(s) to the K8s docs?

Author: I went with the other suggestion you made and added a before you begin section.

- [Horizontal Pod Autoscaling](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/)
- [HorizontalPodAutoscaler Walkthrough](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/)
 
# Horizontal Pod Autoscaling in CD-as-a-Service

> Continuous Deployment-as-a-Service (CD-as-a-Service) converts `HorizontalPodAutoscaler` configured deployments to replica sets. The `HorizontalPodAutoscaler` configuration file is automatically re-written to reference the generated `ReplicaSet` resource. CD-as-a-Service freezes scaling behavior during a deployment.

Reviewer 2: ReplicaSet / ReplicaSets  should not be replica sets

> Continuous Deployment-as-a-Service (CD-as-a-Service) converts `HorizontalPodAutoscaler` configured deployments to replica sets. The `HorizontalPodAutoscaler` configuration file is automatically re-written to reference the generated `ReplicaSet` resource. CD-as-a-Service freezes scaling behavior during a deployment.

Reviewer: I thought we talked about keeping Dan's original wording here??

{{% alert title="Warning" color="warning" %}}You must declare your HPA in a manifest file.{{% /alert %}}

For example, consider a deployment to upgrade the v1 version of a CD-as-a-Service app to a v2 release. In this scenario the v1 application is running with 10 replicas and is scaled with a minimum set to 5 pods and a maximum set to 15 pods.

> 1. When the deployment begins the v1 `HorizontalPodAutoscaler` is deleted. The v1 app is running 10 pods, but no longer scales between 5 and 15 pods.

Reviewer: ```suggestion
1. When the deployment begins, the v1 `HorizontalPodAutoscaler` is deleted. The v1 app is running 10 pods, but no longer scales between 5 and 15 pods.
```

2. CD-as-a-Service deploys the v2 instance of the application with 10 pods (20 total pods between v1 and v2).
3. At the end of the deployment, a new `HorizontalPodAutoScaler` resource is created to target the v2 application. 
4. If you need to roll back your deployment, CD-as-a-Service deletes the v1 app and recreates the original `HorizontalPodAutoscaler`.

> To check the current status of the `HorizontalPodAutoscaler` run `kubectl get hpa`.

## What happens during a rollback

> During a rollback, CD-as-a-Service relies on the `kubectl.kubernetes.io/last-applied-configuration` annotation to re-create the `HorizontalPodAutoScaler`. 

Reviewer: ```suggestion
During a rollback, CD-as-a-Service relies on the `kubectl.kubernetes.io/last-applied-configuration` annotation to recreate the `HorizontalPodAutoScaler`. 
```

> Do not delete this annotation between deployments. CD-as-a-service may not capture any manual or programmatic changes you make to the `HorizontalPodAutoscaler` configuration file between deployment.

> 

Reviewer: Where is the section on "Where to declare your HPA configuration" ?? I think the user needs to know they must declare the `HorizontalPodAutoscaler` in a manifest file rather than creating it using `kubectl`.  https://armory.slab.com/posts/how-c-daa-s-implements-kubernetes-horizontal-pod-autoscaling-egtl4pm5#horfq-where-to-declare-your-hpa-configuration

---

> title: "Kubernetes HPA deployments in CD-as-a-Service"

Reviewer: ```suggestion
title: "Kubernetes HPA Deployments in CD-as-a-Service"
```
Titles use Capital case ([style guide](https://docs.armory.io/contribute/create/guides/docs-style-guide/#use-capital-case-for-page-titles)

linktitle: "Kubernetes HPA"
description: >
  Learn how Armory Continuous Deployment-as-a-Service supports Kubernetes deployments with Horizontal Pod Autoscaling.
---
 # What is HPA?

>  Kubernetes Horizontal Pod Autoscaling (`HPA`) scales deployments based on workload. When the load increases, new pods are generated on the deployed cluster. With `HPA`, the cluster is scaled horizontally by increasing the number of deployed pods on the cluster, rather than vertically increasing memory or CPU of an existing pod. When the workload decreases the cluster is scaled back down to the configured minimum number of replicas as defined in the deployment file.

Reviewer: A few proofreading nits
punctuation, active vs passive
```suggestion
 Kubernetes Horizontal Pod Autoscaling (HPA) scales deployments based on workload. When the load increases, Kubernetes generates new pods on the deployed cluster. With HPA, Kubernetes scales the cluster horizontally by increasing the number of deployed pods on the cluster, rather than vertically increasing the memory or CPU of an existing pod. When the workload decreases, Kubernetes scales the cluster down to the configured minimum number of replicas as defined in the deployment file.
```

Reviewer 2: > With HPA, Kubernetes scales the cluster horizontally

With HPA, Kubernetes scales the application horizontally

>When the workload decreases, Kubernetes scales the cluster down

When the workload decreases, Kubernetes scales the application down

>  - `HorizontalPodAutoscaler` controls the scale of a deployment and its `ReplicaSet`. 

Reviewer: This sentence is verbatim from the Kubernetes doc on Horizonal Pod Autoscaling.

 - The minimum and maximum number of replicas is defined in the deployment file. 
 - The default sync interval is 30 seconds. The sync interval is set by the `kube-controller-manager`. 
 - The default interval is modified using the `horizontal-pod-autoscaler-sync-period` flag. 
 - The default delay is 3 minutes after an upscale event to allow metrics to stabilize. Change the default upscale delay period using the `horizontal-pod-autoscaler-upscale-delay` flag.
 - With `HPA` configured, the deployment object manages the size of underlying replica sets. `HPA` does not manage rolling updates by manipulating the replication controller. 

>   > `HPA` is not recommended for replication controllers. 

Reviewer: Back to my original comment about not really needing to repeat the content of the Kubernetes [Horizontal Pod Autoscaling doc](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/). 

As a general rule, we don't repeat what's in third party documentation or tell people how to use third party products. How are lines 11-16 relevant to CDaaS?

The subject of this page is how CD-as-a-Service implements Kubernetes Horizontal Pod Autoscaling. I think the expectation is that the reader should be familiar with K8s HPA before reading this page's content.

To learn more about `HPA`, read these Kubernetes Documentation topics: 
- [Horizontal Pod Autoscaling](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/)

> - [HorizontalPodAutoscaler Walkthrough](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/)

Reviewer: How about putting this in a **Before you begin** section similar to what's in the [Slab doc](https://armory.slab.com/posts/how-c-daa-s-implements-kubernetes-horizontal-pod-autoscaling-egtl4pm5)?

 
# HPA in CD-as-a-Service

> Continuous Deployment-as-a-Service (CD-as-a-Service) converts `HPA` configured deployments to replica sets. The `HPA` configuration file is automatically re-written to reference the generated `ReplicaSet` resource. CD-as-a-Service freezes scaling behavior during a deployment.

Reviewer: ```suggestion
When you use CD-as-a-Service to deploy a `Deployment` resource,  CD-as-a-Service converts the `Deployment` to a `ReplicaSet`. Correspondingly, when you provide an `HorizontalPodAutoscaler` to scale a `Deployment` resource, that `HorizontalPodAutoscaler` is automatically rewritten to reference the generated `ReplicaSet` resource. CD-as-a-Service freezes scaling behavior during a deployment. 
```
`Depolyment`, `ReplicaSet`, and `HorizontalPodAutoscaler` are Kubernetes `kind` objects and that `kind` is specified in a Kubernetes manifest.

I created an example manifests and deployment file in a  [Slab doc](https://armory.slab.com/posts/hpa-83crfhl7) to illustrate Dan's summary of what CDaaS does with a `Deployment` and corresponding `HorizontalPodAutoscaler`.

> > To check the current status of the `HorizontalPodAutoscaler` run `kubectl get hpa`.

Reviewer: This line seems out of place in this location.

---

> title: "How CD-as-a-Service implements Kubernetes HPA"

Reviewer: ```suggestion
title: "How CD-as-a-Service Implements Kubernetes HPA"
```

Author: I changed the title during self review - if you don't like it I am OK with going with this suggestion - please review updates.

> linktitle: "HPA"

Reviewer: ```suggestion
linktitle: "Kubernetes HPA"
```

description: >

>   Armory CD-as-a-Service supports Kubernetes deployments with HPA.

Reviewer: ```suggestion
  Learn how Armory Continuous Deployment-as-a-Service supports Kubernetes deployments with Horizontal Pod Autoscaling.
```

For SEO, to capture variations of key subjects in title and description, I like to use CD-as-a-Service in one spot and Continuous Deployment-as-a-Service in the other

Author: I made some changes that do not allow me to commit this change - I will make it as my own commit because i like it and struggled with how to SD this topic.

---
 ## What is HPA?

>  Kubernetes Horizontal Pod Autoscaling (HPA) scales deployments in reaction to changes in observed metrics. When the deployment load increases, the number of deployed pods is increased to meet the demand. The scale behavior is horizontal, rather than vertical, which means that rather than assigning an existing pod more memory or CPU to handle the incoming traffic, new pods that identical instances of your application, are created.

Reviewer: ```suggestion
 Kubernetes Horizontal Pod Autoscaling (HPA) scales deployments in reaction to changes in observed metrics. When the deployment load increases, the number of deployed pods is increased to meet the demand. The scale behavior is horizontal, rather than vertical.  HPA scales a resource by adding or removing identical instances of your app instead of Kubernetes assigning an existing pod more resources to handle the incoming traffic.
```

Author: Please see rewrite

 An HPA resource has three components:
- A set of one or more metrics, each with a desired target value.
- A reference to a Kubernetes resource, typically a Deployment but any resource that supports the Scale API sub-resource (e.g., `ReplicaSet`, `StatefulSet`). The HPA scales this resource in response to changes in observed metrics.
- A minimum and maximum number of replicas.

HPA checks metrics as configured in the deployment file AT A DEFAULT 30 SEC intervals. When the threshold load is met the HPA increases or decreases the number of pods. 
 
 Kubernetes implements Horizontal Auto Scaling (HPA) as a control loop that runs intermittently. `HorizontalPodAutoscaler` controls the scale of a Deployment and its `ReplicaSet`. 
 
 HPA is not recommended for replication controllers. 
 
 HPA does not manage rolling updates by manipulating the replication controller. The deployment object manages the size of underlying replica sets. 

 The sync interval is set by the `kube-controller-manager`. The default interval is 30 seconds. Change the default interval using the `horizontal-pod-autoscaler-sync-period` flag.

> The default delay is 3 minutes after an upscale event to allow metrics to stabilize. Change the default upscale delay period using the `horizontal-pod-autoscaler-upscale-delay` flag.

Reviewer: I'm not sure this "What is HPA" section is even needed since it basically restates (or almost copies verbatim) what's in the Kuberntes docs. What about replacing section with a  prereq section that points to the relevant K8s docs?

Author: Please see rewrite

 
## How HPA works in CD-as-a-Service

> When HPA is configured for an Armory Continuous Deployment-as-a-Service (CD-as-a-Service) deployment the deployment is converted to a `ReplicaSet` and the HPA configuration file is automatically re-written to reference the generated `ReplicaSet` resource. Continuous Deployment-as-a-Service freezes scaling behavior during a deployment.

Reviewer: ```suggestion
When you use CD-as-a-Service to deploy a `Deployment` resource,  CD-as-a-Service converts the `Deployment` to a `ReplicaSet`. Correspondingly, when you provide an `HPA` to scale a `Deployment` resource, that `HPA` is automatically rewritten to reference the generated `ReplicaSet` resource.
```

Author: please see rewrite

> Consider a deployment to upgrade from the v1 to v2 version of your application. In this scenario the v1 application is running with 10 replicas and is scaled by HPA (min: 5 pods, max: 15 pods).

Reviewer: ```suggestion
CD-as-a-Service freezes scaling behavior during a deployment. Consider a deployment to upgrade from the `v1` to `v2` version of your app. In this scenario, the `v1` app is running with 10 replicas and is scaled by `HPA` (`min: 5 pods`, `max: 15 pods`).
```

Author: Please see rewrite

>  - At the beginning of a deployment, the HPA is deleted. Your v1 application has 10 pods but no longer scales between 5 and 15 pods.

Reviewer: ```suggestion
 - At the beginning of a deployment, CD-as-a-Service deletes the `HPA`. Your `v1` app has 10 pods but no longer scales between 5 and 15 pods.
```

Author: please see rewrite

>  - CD as a Service deploys the v2 instance of your application with 10 pods (20 total pods between v1 and v2).

Reviewer: ```suggestion
 - CD-as-a-Service deploys the `v2` instance of your application with 10 pods (20 total pods between `v1` and `v2`).
```

Author: fixed the product name consistency throughout during re-write

Author: will codify v1/v2 in new commit

>  - At the end of the deployment, a new HPA resource is created to target the v2 application.

Reviewer: ```suggestion
 - At the end of the deployment, CD-as-a-Service creates a new `HPA` resource to target the `v2` app.
```

Author: please see rewrite

>  - If the deployment needs to be rolled back, CDaaS deletes the v2 application and re-creates the original HPA.

Reviewer: ```suggestion
 - If you need to roll back your deployment, CD-as-a-Service deletes the `v2` app and recreates the original `HPA`.
```

Author: fixed in new commit

### Supported versions

> Armory Continuous Deployment-as-a-Service supports the following HorizontalPodAutoscaler API versions:

Reviewer: ```suggestion
CD-as-a-Service supports the following `HorizontalPodAutoscaler` API versions:
```

Author: please see rewrite - the fix is there but I cannot commit suggestion.

- [v1](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/horizontal-pod-autoscaler-v1/)
- [v2beta 1 and 2](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/horizontal-pod-autoscaler-v2beta2/)

> -  [v2](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/horizontal-pod-autoscaler-v2/)

Reviewer: Great that you added the links. What do you think about moving Supported Versions section to a "##" near or at the bottom?

Author: I did this in my re-write - totally agree

  
**Before deployment**

- Upgrade a database schema
- Custom approval process

**Within your deployment strategy**

- Check logs and system health

> - Run custom tests

Reviewer: Curious why these lines are here? Looks like maybe you copied from the Webhook-Based Approvals page and forgot to remove?

**After deployment**

> Once CPU utilization drops to 0, the HPA automatically scales the number of replicas down to 1.

Reviewer: Not sure why this is here?

Author: Removed :)

> Autoscaling the replicas may take a few minutes.

Run integration tests in a staging environment.
- Perform metric tests

> - Run security scanners

Reviewer: Again wondering why these lines are here.

Author: removed

**Check the current status of the `HorizontalPodAutoscaler`**
```
kubectl get hpa
```

> ### Roll back an HPA configured deployment

Reviewer: ```suggestion
### What happens during a rollback
```

> During a rollback, Armory CDaaS relies on the `kubectl.kubernetes.io/last-applied-configuration` annotation to re-create the HPA. 

Reviewer: ```suggestion
During a rollback, CD-as-a-Service relies on the `kubectl.kubernetes.io/last-applied-configuration` annotation to recreate the `HPA`. 
```

Author: I updated this without seeing your comment - please re-review :)

> Do not delete this annotation between deployments. 

> Manual or programmatic changes to the HPA configuration file between deployments (for example, changes to the HPA's min/max boundaries) may not be captured during a rollback.

Reviewer: ```suggestion
CD-as-a-service may not capture any manual or programmatic changes you make to the `HPA` configuration file between deployment.
```

Author: Again the rewrite I did over the weekend does not allow the commit please re-review.

> Point to an HPA configuration file that references a Deployment resource in your manifests. An HPA configuration file can be assigned to each individual Deployment resource.

Reviewer: Unsure why this paragraph is in the section about the known rollback issue.

What about adding a separate `##` section 
```
## Where to declare your HPA configuration
CD-as-a-Service requires your app's `HPA` be declared in your app's manifest. Provide an `HPA` that references a `Deployment` resource in that manifest. Each `Deployment` resource can have its own `HPA`.
```

Author: Please see rewrite

Reviewer: I don't see how/where to declare your HPA configuration anywhere in the doc. A k8s user can create an HPA using kubectl or declaratively in a manifest. [CDaaS requires an HPA manifest](https://armory.slab.com/posts/how-c-daa-s-implements-kubernetes-horizontal-pod-autoscaling-egtl4pm5#horfq-where-to-declare-your-hpa-configuration).

```
## Where to declare your HPA configuration

CD-as-a-Service requires your app's HPA be [declared in a manifest](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/#creating-the-autoscaler-declaratively).  Each `Deployment` resource can have its own `HPA`.
```
